### PR TITLE
Continuous fuzzing

### PR DESCRIPTION
**Pull request by bookmoons**
Proposing to integrate with [Fuzzit](https://fuzzit.dev/) for automated bug discovery.

This fuzzes code generation and unmarshaling of some generated code. It's complementary to nice existing fuzzing of marshaling. For unmarshaling I generated from the nice test struct in the existing fuzzing.

If you wanted to get marshaling running on the platform too I think it's possible to link up by fuzzing the entropy used to populate objects. If you're interested I can look into it.

There's a [successful run](https://travis-ci.org/bookmoons/ffjson/builds/577084126) under my Travis account. The PR build will fail due to missing an API key. If you're interested in trying it setup is like this:
* In Fuzzit create targets `ffjson-generate` `ffjson-unmarshal`.
* In Fuzzit settings grab an API key. In repo settings in [Travis](https://travis-ci.org/) paste it to envvar `FUZZIT_API_KEY`.

### CODE REVIEW

**Comment by erikdubbelboer** [comment]
If you commit `tests/fuzz/generated.go` every pull request will have to update it as well otherwise it would never fuzz the latest version. It's much better if this file is automatically generated only when the fuzzer is build.

File: tests/fuzz/target_fuzz.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+package fuzz
+
+// FuzzUnmarshal tests unmarshaling
+func FuzzUnmarshal(fuzz []byte) int {
+	data := &Data{}
+	err := data.UnmarshalJSON(fuzz)
+	if err != nil {
+		return 0
+	}
```
Comment (erikdubbelboer) [inline tests/fuzz/target_fuzz.go]: I would also Marshal the result back to `[]byte` here to kinda test that as well.

File: .travis.yml [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
 
+jobs:
+  include:
+    - stage: Fuzz regression
+      go: 1.12.x
+      dist: bionic
+      script: ./fuzzit.sh local-regression
+
+    - stage: Fuzz
+      if: branch = master AND type IN (push)
```
Comment (erikdubbelboer) [inline .travis.yml]: This does not do what you think it does. From the travis documentation:
```
require the branch name to be master (note for PRs this is the base branch name)
```
So this doesn't just run on pushes to master. It also runs to pushes to pull requests that want to be merged into master (so master is their base branch).

**Comment by erikdubbelboer** [comment]
The problem I had with setting it in the repo settings is that it won't be set for pull requests which prevented me from running local-regression tests in pull requests.

File: .travis.yml (hunk 2 above)
Comment (erikdubbelboer) [reply .travis.yml]: I'm doing it like this: https://github.com/valyala/fasthttp/blob/master/.travis.yml#L56-L57

File: fuzzit.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+# Setup
+make
+(
+    cd tests/fuzz
+    rm -f target_ffjson.go
+    ffjson target.go
+)
+export GO111MODULE="off"
+go get -u github.com/dvyukov/go-fuzz/go-fuzz github.com/dvyukov/go-fuzz/go-fuzz-build
```
Comment (pquerna) [inline fuzzit.sh]: I don't love using `go get` like this to install go-fuzz, or the `./...` bellow this.  Is there a way to get a validated release of these tools or a base docker image we can more cleanly depend upon?

Comment (bookmoons) [reply fuzzit.sh]: Will investigate this.

**Comment by bookmoons** [comment]
Thanks for looking at it @pquerna.

**Comment by bookmoons** [comment]
Thanks a lot for reviewing @pquerna. I think you can remove me from the org in settings, then it's all yours.

**Comment by bookmoons** [comment]
Oh, I see the fuzz stage failed. Is it possible you used an API key from your default org? They give different keys to the different orgs. If that was it, you can switch org in the drop down then go to settings to get the right key.

**Pull request merged**